File: bhnet.py
#!/usr/bin/python3

import sys
import threading
import subprocess
import getopt
import socket
from colorama import Style, Fore

listen = False
command = False
upload = False
execute = ""
target = ""
upload_destination = ""
port = 0

def usage():
    print(Style.BRIGHT + Fore.CYAN + ''' 
 ________      ________    ___  ___     
|\   __  \    |\   __  \  |\  \|\  \    
\ \  \|\ /_   \ \  \|\  \ \ \  \_\  \   
 \ \   __  \   \ \   ____\ \ \   __  \    Net Tool
  \ \  \|\  \   \ \  \___|  \ \  \ \  \ 
   \ \_______\   \ \__\      \ \__\ \__\      
    \|_______|    \|__|       \|__|\|__|
                                        
                Recreated By Byteninja9
    ''')
    print("Usage: bhpnet.py -t target_host -p port")
    print('''-l --listen                         - listen  on [host]:[port] for 
                                                   incoming connections''')
    print('''-e --execute=file_to_run            - execute the given file upon  
                                                   receiving connections''')
    print('''-c --command                        - initialize a command shell''')
    print('''-u --upload=destination             - upon receiving connection upload a 
                                                   file and write to [destination]''')
    print()

    print(Style.BRIGHT + Fore.GREEN + "Examples: ")
    print("bhpnet.py -t 192.168.0.1 -p 5555 -l -c")
    print("bhpnet.py -t 192.168.0.1 -p 5555 -l -c -u=c:\\target.exe")
    print("bhpnet.py -t 192.168.0.1 -p 5555 -l -e=\"cat /etc/passwd\"")
    print("echo 'ABCDEFGH' | ./bhpnet.py -t 192.168.0.1 -p 135")
    sys.exit(0)

def client_sender(buffer):
    client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        client.connect((target, port))

        if len(buffer):
            client.send(buffer.encode())
        
        while True:
            recv_len = 1
            res = ""

            while recv_len:
                data = client.recv(4096)
                recv_len = len(data)
                res += data.decode()

                if recv_len < 4096:
                    break
            
            print(res)
            buffer = input(b"").strip()
            buffer += "\n"

            client.send(buffer.encode())

    except Exception as e:
        print(Style.BRIGHT + Fore.RED + f"[+] Exception: {e}. Exiting." + Fore.RESET)
        client.close()

def client_handler(client_socket):
    global upload
    global execute
    global command

    print("[+] Client connected")

    if len(upload_destination):
        file_buffer = b""

        while True:
            data = client_socket.recv(1024)
            if not data:
                break
            else:
                file_buffer += data

        try:
            with open(upload_destination, 'wb') as file_descriptor:
                file_descriptor.write(file_buffer)
            client_socket.send((Style.BRIGHT + Fore.GREEN + f"Successfully saved to {upload_destination}\r\n" + Fore.RESET).encode())

        except Exception as e:
            client_socket.send((Style.BRIGHT + Fore.RED + f"Failed to save file to {upload_destination}: {e}").encode())

    if len(execute):
        output = run_command(execute)
        client_socket.send(output.encode())

    if command:
        while True:
            try:
                client_socket.send((Style.BRIGHT + Fore.LIGHTCYAN_EX + "<BHP:#> " +  Fore.RESET).encode())
                cmd_buffer = b""
                while b"\n" not in cmd_buffer:
                    data = client_socket.recv(1024)
                    if not data:
                        break
                    cmd_buffer += data

                cmd_buffer = cmd_buffer.decode().strip()
                print(f"[+] Received command: {cmd_buffer}")
                
                if cmd_buffer:
                    res = run_command(cmd_buffer)
                    print(f"[+] Command output: {res}")
                    client_socket.send(res.encode())
                else:
                    client_socket.send((Style.BRIGHT + Fore.RED + "No command received\r\n" + Fore.RESET).encode())

            except Exception as e:
                print(f"[+] Exception in command handling: {e}")
                client_socket.close()
                break


def server_loop():
    global target

    if not len(target):
        target = "0.0.0.0"

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind((target, port))
    server.listen(5)
    print(f"[+] Listening on {target}:{port}")

    while True:
        client_socket, addr = server.accept()
        print(f"[+] Accepted connection from {addr}")
        client_thread = threading.Thread(target=client_handler, args=(client_socket,))
        client_thread.start()

def run_command(command):
    command = command.rstrip()

    try:
        output = subprocess.check_output(command, stderr=subprocess.STDOUT, shell=True)
    except subprocess.CalledProcessError as e:
        output = f"Failed to execute command: {e}\r\n"

    return output.decode()

def main():
    global listen
    global port
    global execute
    global target
    global upload_destination

    if not len(sys.argv[1:]):
        usage()

    try:
        opts, args = getopt.getopt(sys.argv[1:], "hle:t:p:cu", ["help", "listen", "execute=", "target=", "port=", "command", "upload="])
    
    except getopt.GetoptError as err:
        print(Style.BRIGHT + Fore.RED + str(err) + Fore.RESET)
        usage()

    for o, a in opts:
        if o in ("-h", "--help"):
            usage()
        elif o in ("-l", "--listen"):
            listen = True 
        elif o in ("-e", "--execute"):
            execute = a
        elif o in ("-c", "--command"):
            command = True
        elif o in ("-u", "--upload"):
            upload_destination = a
        elif o in ("-t", "--target"):
            target = a
        elif o in ("-p", "--port"):
            port = int(a)
        else:
            assert False, "Invalid Option"
    
    if not listen and len(target) and port > 0:
        buffer = sys.stdin.read()
        client_sender(buffer)

    if listen:
        server_loop()

if __name__ == "__main__":
    main()

File: bhnet.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>

#define BUFFER_SIZE 4096

// ANSI escape codes for colors
#define RESET "\033[0m"
#define BRIGHT "\033[1m"
#define CYAN "\033[36m"
#define GREEN "\033[32m"
#define RED "\033[31m"
#define LIGHTCYAN "\033[96m"

// Function prototypes
void print_usage();
void execute_command(int client_socket);
void upload_file(int client_socket, const char *upload_destination);
void run_file(int client_socket, const char *file_to_run);
void server_loop(int port, const char *upload_destination, const char *file_to_run, int command_mode);

// Global variables for server options
int command_mode = 0;
char *upload_destination = NULL;
char *file_to_run = NULL;
char *target = NULL;

void print_usage() {
    printf(BRIGHT CYAN "\n"
           " ________      ________    ___  ___     \n"
           "|\\   __  \\    |\\   __  \\  |\\  \\|\\  \\    \n"
           "\\ \\  \\|\\  \\   \\ \\  \\|\\  \\ \\ \\  \\_\\  \\   \n"
           " \\ \\  \\\\\\  \\   \\ \\   __  \\ \\ \\   __  \\   Net Tool\n"
           "  \\ \\  \\\\\\  \\   \\ \\  \\ \\  \\ \\ \\  \\ \\  \\ \n"
           "   \\ \\_______\\   \\ \\__\\ \\__\\ \\ \\__\\ \\__\\ \n"
           "    \\|_______|    \\|__|\\|__|  \\|__|\\|__| \n"
           "                                \n"
           "                Recreated By Byteninja9\n"
           RESET);

    printf("Usage: net_tool -p port [options]\n");
    printf("-p --port                           - port to listen on\n");
    printf("-e --execute=file_to_run             - execute the given file upon receiving connection\n");
    printf("-c --command                        - initialize a command shell\n");
    printf("-u --upload=destination             - upon receiving connection upload a file and write to [destination]\n");
    printf("-t --target=host                    - target host\n");
    printf("-h --help                           - show this help message\n");
    printf("\n");
    printf(BRIGHT GREEN "Examples:\n");
    printf("net_tool -p 5555 -c\n");
    printf("net_tool -p 5555 -c -u=/path/to/upload/file\n");
    printf("net_tool -p 5555 -e=\"/path/to/script.sh\"\n");
    printf("net_tool -p 5555 -t 192.168.0.1\n");
    printf(RESET);
    exit(0);
}

void execute_command(int client_socket) {
    char buffer[BUFFER_SIZE];
    int n;

    while (1) {
        // Send the command prompt
        const char *prompt = BRIGHT LIGHTCYAN "<BHP:#> " RESET;
        send(client_socket, prompt, strlen(prompt), 0);

        // Read the command from the client
        memset(buffer, 0, BUFFER_SIZE);
        n = recv(client_socket, buffer, BUFFER_SIZE - 1, 0);
        if (n <= 0) {
            perror("recv");
            break;
        }

        // Remove newline character at the end
        buffer[strcspn(buffer, "\n")] = 0;

        printf(BRIGHT CYAN "[+] Received command: %s" RESET "\n", buffer);

        // Execute the command
        FILE *fp;
        char output[BUFFER_SIZE];
        memset(output, 0, BUFFER_SIZE);

        fp = popen(buffer, "r");
        if (fp == NULL) {
            perror("popen");
            send(client_socket, (RED "Failed to execute command.\n" RESET), 27, 0);
            continue;
        }

        // Send the command output to the client
        while (fgets(output, sizeof(output), fp) != NULL) {
            send(client_socket, output, strlen(output), 0);
        }

        printf(BRIGHT GREEN "[+] Command output sent." RESET "\n");

        pclose(fp);
    }

    // Close the client socket
    close(client_socket);
}

void upload_file(int client_socket, const char *upload_destination) {
    char buffer[BUFFER_SIZE];
    int n;

    FILE *file = fopen(upload_destination, "wb");
    if (file == NULL) {
        perror("fopen");
        send(client_socket, (RED "Failed to open file for writing.\n" RESET), 34, 0);
        return;
    }

    // Receive the file from the client
    while ((n = recv(client_socket, buffer, BUFFER_SIZE, 0)) > 0) {
        fwrite(buffer, 1, n, file);
    }

    if (n < 0) {
        perror("recv");
        send(client_socket, (RED "Failed to receive file.\n" RESET), 25, 0);
    } else {
        printf(BRIGHT GREEN "[+] Successfully saved file to %s\n" RESET, upload_destination);
        send(client_socket, (GREEN "File successfully uploaded.\n" RESET), 26, 0);
    }

    fclose(file);
}

void run_file(int client_socket, const char *file_to_run) {
    char output[BUFFER_SIZE];

    // Execute the file
    FILE *fp = popen(file_to_run, "r");
    if (fp == NULL) {
        perror("popen");
        send(client_socket, (RED "Failed to execute file.\n" RESET), 25, 0);
        return;
    }

    // Send the file output to the client
    while (fgets(output, sizeof(output), fp) != NULL) {
        send(client_socket, output, strlen(output), 0);
    }

    pclose(fp);
}

void server_loop(int port, const char *upload_destination, const char *file_to_run, int command_mode) {
    int server_socket, client_socket;
    struct sockaddr_in server_addr, client_addr;
    socklen_t client_len;

    // Create a socket
    server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket == -1) {
        perror("socket");
        exit(EXIT_FAILURE);
    }

    // Set up the server address struct
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = INADDR_ANY;
    server_addr.sin_port = htons(port);

    // Bind the socket to the specified port
    if (bind(server_socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) == -1) {
        perror("bind");
        close(server_socket);
        exit(EXIT_FAILURE);
    }

    // Listen for incoming connections
    if (listen(server_socket, 5) == -1) {
        perror("listen");
        close(server_socket);
        exit(EXIT_FAILURE);
    }

    printf(BRIGHT GREEN "[+] Listening on port %d...\n" RESET, port);

    // Main server loop
    while (1) {
        client_len = sizeof(client_addr);
        client_socket = accept(server_socket, (struct sockaddr *)&client_addr, &client_len);
        if (client_socket == -1) {
            perror("accept");
            close(server_socket);
            exit(EXIT_FAILURE);
        }

        printf(BRIGHT GREEN "[+] Accepted connection from %s:%d\n" RESET,
               inet_ntoa(client_addr.sin_addr), ntohs(client_addr.sin_port));

        // Handle client request based on mode
        if (upload_destination) {
            upload_file(client_socket, upload_destination);
        } else if (file_to_run) {
            run_file(client_socket, file_to_run);
        } else if (command_mode) {
            execute_command(client_socket);
        }

        // Close the client socket
        close(client_socket);
    }

    // Close the server socket
    close(server_socket);
}

int main(int argc, char *argv[]) {
    int port = -1;
    char *endptr;

    // Default values for options
    upload_destination = NULL;
    file_to_run = NULL;
    command_mode = 0;

    if (argc < 3 || strcmp(argv[1], "-p") != 0) {
        print_usage();
    }

    port = strtol(argv[2], &endptr, 10);
    if (*endptr != '\0' || port <= 0 || port > 65535) {
        fprintf(stderr, RED "Invalid port number: %s\n" RESET, argv[2]);
        exit(EXIT_FAILURE);
    }

    // Parse additional options
    for (int i = 3; i < argc; i++) {
        if (strcmp(argv[i], "-c") == 0) {
            command_mode = 1;
        } else if (strncmp(argv[i], "-u=", 3) == 0) {
            upload_destination = argv[i] + 3;
        } else if (strncmp(argv[i], "-e=", 3) == 0) {
            file_to_run = argv[i] + 3;
        } else if (strncmp(argv[i], "-t=", 3) == 0) {
            target = argv[i] + 3;
        } else {
            print_usage();
        }
    }

    // Validate options
    if (command_mode && (upload_destination || file_to_run)) {
        fprintf(stderr, RED "Command mode (-c) cannot be used with upload or execute options.\n" RESET);
        exit(EXIT_FAILURE);
    }

    if (upload_destination && file_to_run) {
        fprintf(stderr, RED "Cannot use both upload (-u) and execute (-e) options simultaneously.\n" RESET);
        exit(EXIT_FAILURE);
    }

    if (command_mode || upload_destination || file_to_run) {
        server_loop(port, upload_destination, file_to_run, command_mode);
    } else {
        print_usage();
    }

    return 0;
}
